File: built-in plugins dsp/mb-filter-comb.cpp
/**
 * MB Comb Filter
 * Category : effect
 * Type     : eq
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Comb filter for metallic and tuned resonance effects
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_FILTER_COMB_H
#define MB_FILTER_COMB_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbFilterComb : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-filter-comb";
    static constexpr const char* PLUGIN_NAME    = "MB Comb Filter";
    static constexpr const char* PLUGIN_TYPE    = "eq";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float freq = 500f;  // range [50, 5000]
    float feedback = 0.5f;  // range [-0.99, 0.99]
    float damping = 0.5f;  // range [0, 1]
    float mix = 0.5f;  // range [0, 1]
    };

    MbFilterComb() = default;
    ~MbFilterComb() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.freq = std::clamp(params.freq, 50f, 5000f);
        params.feedback = std::clamp(params.feedback, -0.99f, 0.99f);
        params.damping = std::clamp(params.damping, 0f, 1f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Comb Filter
        return input;
    }
};

#endif // MB_FILTER_COMB_H

File: maxbooster7.5/built-in plugins dsp/mb-filter-comb.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbFilterCombPlugin: PluginDefinition = {
  id: "mb-filter-comb",
  slug: "mb-filter-comb",
  name: "MB Comb Filter",
  category: "effect",
  type: "eq" as any,
  version: "1.0.0",
  description: "Comb filter for metallic and tuned resonance effects",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "freq",
      name: "Frequency",
      type: "float",
      defaultValue: 500,
      minValue: 50,
      maxValue: 5000,
      automatable: true,
    },
    {
      id: "feedback",
      name: "Feedback",
      type: "float",
      defaultValue: 0.5,
      minValue: -0.99,
      maxValue: 0.99,
      automatable: true,
    },
    {
      id: "damping",
      name: "Damping",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "mix",
      name: "Mix",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { freq: 500, feedback: 0.5, damping: 0.5, mix: 0.5 },
};

export default MbFilterCombPlugin;
